src/core/tiledmesh/qgstiledmeshindex.h

```diff
@@ -0,0 +1,239 @@
+
+#include "qgis_core.h"
+#include "qgis.h"
+
+#include <QCache>
+#include <QReadWriteLock>
+
+class QgsTiledMeshTile;
+class QgsFeedback;
+class QgsTiledMeshRequest;
+
+#ifndef SIP_RUN
+
+/**
+ * \ingroup core
+ * \brief An abstract base class for tiled mesh data provider indices.
+ *
+ * \warning Subclasses must take care to ensure that their operation is thread safe, as the same
+ * index object will potentially be used by multiple threads concurrently.
+ *
+ * \since QGIS 3.34
+ */
+class CORE_EXPORT QgsAbstractTiledMeshIndex
+{
+  public:
+
+    QgsAbstractTiledMeshIndex();
+    virtual ~QgsAbstractTiledMeshIndex();
+
+    //! QgsAbstractTiledMeshIndex cannot be copied
+    QgsAbstractTiledMeshIndex( const QgsAbstractTiledMeshIndex &other ) = delete;
+    //! QgsAbstractTiledMeshIndex cannot be copied
+    QgsAbstractTiledMeshIndex &operator=( const QgsAbstractTiledMeshIndex &other ) = delete;
+
+    /**
+     * Returns the root tile for the index.
+     */
+    virtual QgsTiledMeshTile rootTile() const = 0;
+
+    /**
+     * Returns the tile ID of the parent tile of the tile with matching \a id, or an empty string
+     * if the tile has no parent.
+     *
+     * \see childTileIds()
+     */
+    virtual QString parentTileId( const QString &id ) const = 0;
+
+    /**
+     * Returns a list of the tile IDs of any children for the tile with matching \a id.
+     *
+     * \see parentTileId()
+     */
+    virtual QStringList childTileIds( const QString &id ) const = 0;
+
+    /**
+     * Returns the tile with matching \a id, or an invalid tile if the matching tile is not available.
+     */
+    virtual QgsTiledMeshTile getTile( const QString &id ) = 0;
+
+    /**
+     * Returns the tile IDs which match the given \a request.
+     *
+     * An optional \a parentID can be used to limit the results to children of a specific tile.
+     *
+     * May return an empty list if no data satisfies the request.
+     */
+    virtual QStringList getTiles( const QgsTiledMeshRequest &request, const QString &parentId = QString() ) = 0;
+
+    /**
+     * Retrieves index content for the specified \a uri.
+     *
+     * The content is cached within the index, so multiple calls are potentially cost-free.
+     *
+     * The optional \a feedback argument can be used the cancel the request early.
+     */
+    QByteArray retrieveContent( const QString &uri, QgsFeedback *feedback = nullptr );
+
+    /**
+     * Returns TRUE if the tile with the given \a id can be further refined.
+     *
+     * \see refineAsync()
+     */
+    virtual bool tileCanRefine( const QString &id ) const = 0;
+
+    /**
+     * Refines the tile with the given \a id by fetching any sub datasets attached to the tile.
+     *
+     * Blocks while the tile refinement is in progress.
+     *
+     * Returns TRUE if the refinement was successful.
+     *
+     * \see tileCanRefine()
+     */
+    virtual bool refineAsync( const QString &id, QgsFeedback *feedback = nullptr ) = 0;
+
```

Comment:
One more note about naming - the term "refine" is being used in two slightly different scenarios:
- tile refinement strategy (add / replace)
- refine tile (to load next bit of hierarchy)

Maybe the second "refine" could be changed to something else? (in point clouds I think we used the term to "fetch hierarchy")

Reply:
Ok, I've gone with fetchHierarchy and `Qgis::TileChildrenAvailability childAvailability( const QString &id )` instead.